src/main/java/com/example/planit/engine/CalendarEngine.java

```diff
@@ -1402,22 +1402,69 @@ public static boolean hasAuthorizeScopesStillValid(String access_token) throws I
         return false; // scope was not found or response wasn't successful
     }
 
-    public DTOstudyPlanResponseToController getUserLatestStudyPlan(String sub) {
+    public DTOstudyPlanAndupComingStudySessionResponseToController getUserLatestStudyPlan(String sub) {
 
         try {
 
             Optional<User> maybeUser = userRepo.findUserBySubjectID(sub);
 
             if (maybeUser.isEmpty()) {
-                return new DTOstudyPlanResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
+                return new DTOstudyPlanAndupComingStudySessionResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
             }
-            return new DTOstudyPlanResponseToController(true, NO_PROBLEM, HttpStatus.OK, maybeUser.get().getLatestStudyPlan());
+            User user = maybeUser.get();
+            // refresh access_token before making api call
+            validateAccessTokenExpireTime(user);
+
+            // check if access_token still have scope for Google calendar
+            if (!hasAuthorizeScopesStillValid(user.getAuth().getAccessToken())) {
+                return new DTOstudyPlanAndupComingStudySessionResponseToController(false, Constants.ERROR_NO_VALID_ACCESS_TOKEN, HttpStatus.UNAUTHORIZED);
+            }
+
+            return new DTOstudyPlanAndupComingStudySessionResponseToController(true, NO_PROBLEM, HttpStatus.OK, maybeUser.get().getLatestStudyPlan(), getUpComingSessionForUser(user));
         } catch (Exception e) {
             logger.error(buildExceptionMessage(e));
-            return new DTOstudyPlanResponseToController(false, ERROR_DEFAULT, HttpStatus.INTERNAL_SERVER_ERROR);
+            return new DTOstudyPlanAndupComingStudySessionResponseToController(false, ERROR_DEFAULT, HttpStatus.INTERNAL_SERVER_ERROR);
         }
     }
 
+    private StudySession getUpComingSessionForUser(User user) throws GeneralSecurityException, IOException {
+
+        Calendar calendarService = getCalendarService(user.getAuth().getAccessToken(), user.getAuth().getExpireTimeInMilliseconds());
+
+        List<CalendarListEntry> calendarList = getCalendarList(calendarService);
+
+        StudySession upComingSession = null;
+
+        Instant currentTime = Instant.now();
+        DateTime startTime = new DateTime(currentTime.toEpochMilli());
+
+        int countEvents = 0;
+        Events events = null;
+
+        for (CalendarListEntry calendar : calendarList) {
+            if (calendar.getId().equals(user.getPlanItCalendarID())) {
+                events = calendarService.events().list(calendar.getId())
+                        .setTimeMin(startTime)
+                        .setOrderBy("startTime")
+                        .setSingleEvents(true)
+                        .execute();
```

Comment:
@oshriFadlon1 I suggest you add a "break" command to the if statment (no need to continue the for loop once you've found the PlanIT calendar).

Reply:
add break